FhirToDataLake/docs/functional-test/Native-Parquet.md

```diff
@@ -0,0 +1,95 @@
+
+
+Target:
+1. Empty input
+2. Invalid input types
+3. Large input
+4. Empty result
+
+## Cpp
+
+### Fix
+- Return error code when register parquet schema with empty or whitespace key.
+- Use property key of schema node, rather than "Name" property value to create arrow schema.
+
+    ```
+    ...
+    "resourceType": {
+        "Name": "resourceType",
+        "Type": "string",
+        ...
+        },
+    ...
+    ```
+- Make output error message pointer in ParquetWriter_Write interface can be null, if so will only return error code and not give error message.
+
+    ```
+    int Write(const string& resourceType, const char* inputJson, int inSize, byte** outputData, int* outSize, char* errorMessage=nullptr);
+    ```
+
+### Test
+
+- ParquetLib
+    
+    - int RegisterParquetSchema(ParquetWriter* writer, const char* schemaKey, const char* schemaData);
+       
+       1. Correctly register parquet schema to Parquet writer.
+       2. **[Refine]** Can return error code if schema content is empty, white space or null.
+       3. **[Add]** Can return error code if schema key is empty, white space or null.
+   
+    - int ConvertJsonToParquet(ParquetWriter* writer, const char* schemaKey, const char* inputJson, int inputLength, byte** outputData, int* outputLength, char* errorMessage);
+
+      1. Correctly convert Json data to Parquet.
+      2. **[Add]** Can return error code if schema key is null.
+
+    - int TryReleaseUnmanagedData(byte** data);
+      
+      1. **[Add]** Confirm the data is released.
+
+1. ParquetWriter
+
+    - int RegisterSchema(const string& schemaKey, const string& schemaData);
+      1. Correct register schema with given schema key.
+      2. **[Refine]** Can return error code if schema content is empty, white space, invalid Json or null.
+      3. **[Refine]** Can return error code if schema key is empty, white space or null.
+
+    - int Write(const string& resourceType, const char* inputJson, int inSize, byte** outputData, int* outSize, char* errorMessage=nullptr);
+      1. Correct write single and batch Patient data.
+      2. **[Add]** Can only return error code when error message pointer is null.
+      3. **[Add]** Can give error message and return error code if input data is null.
+      4. **[Add]** Can give error message and return error code if output data pointer or output data size is null.
+
+2. SchemaManager
+    - int AddSchema(const string& schemaKey, const string& schemaJson);
+      1. Correct add schema with given schema key.
+      2. **[Add]** Can return false if schema content is empty, white space or invalid Json.
+      3. **[Add]** Can return false if schema key is empty or white space.
+
+    - shared_ptr<arrow::Schema> GetSchema(const string& schemaKey);
+      1. Correct return schema with given schema key.
+
+## CSharp
```

Comment:
Can we add a work item to track a e2e test if FHIR server has invalid data to transform? like a field with wrong type

Reply:
Sure, create this [item](https://microsofthealth.visualstudio.com/Health/_workitems/edit/97856)